README.md
kgrid
=====

Version 1.2: [Change log](./CHANGELOG.md)

Generates a suitably converged **k**-point grid for solid-state
quantum chemical calculations.

Current status
--------------

Two programs are provided: kgrid and kgrid-series

### kgrid
* The specified input file is read using Atomic Simulation Environment ([supported formats](https://wiki.fysik.dtu.dk/ase/ase/io.html#module-ase.io))
  * If none is specified, looks for geometry.in (FHI-aims) in working directory
* A **k**-point density is selected to satisfy a given length cutoff,
  as described by Moreno & Soler (1992)[1]. The length cutoff
  corresponds to a radius about repeated images that would be needed
  in a gamma-point supercell calculation to achieve the same sampling.
* This **k**-point grid is expressed as a number of samples in each
  lattice vector and passed to standard output. (Note that this is NOT
  a Moreno-Soler grid as it does not use symmetry information to
  minimise the required number of points. It is a uniform grid
  specified with the same length parameter notation.)
* Default **k**-point cutoff is 10Å (generally well-converged for
  semiconducting or insulating materials)
* Optional arguments are implemented with conventional GNU/POSIX
  syntax, including -h help option

### kgrid-series
* Reports a series of "critical" cutoffs and KSPACING values with
  their corresponding k-point grids.
* These may be very useful for convergence testing, but users should be
  wary of rounding behaviour near these values.

Requirements
------------

* Python 3.6
* [Atomic Simulation Environment](https://wiki.fysik.dtu.dk/ase) (ASE) version 3.18

Usage
-----

### kgrid
From the command line

``` bash
    kgrid FILE -t TYPE -c CUTOFF
```

will return a suggested set of mesh dimensions. FILE can be any
[file format supported by ASE](https://wiki.fysik.dtu.dk/ase/ase/io/io.html);
if no FILE is specified, **kgrid** will look for a *geometry.in* file in
the current directory. TYPE is a string specifying the format of this
file; usually this argument can be left out and the correct type will
be inferred by ASE. CUTOFF is the real-space cutoff parameter in Å and
defaults to 10.0.

There is an internal Python function which may prove useful to ASE users, with the form
``` python
kgrid.calc_kpt_tuple(atoms, cutoff_length=10)
```

where `atoms` is an ASE atoms object and the function returns a
tuple. As such, **kgrid** may be used while setting up a calculation
with a typical ASE calculator. For example:

``` python
import kgrid
from ase.io import read
from ase.calculators.vasp import Vasp

atoms = read('my_favourite_structure.cif')
calc = Vasp(xc='PBE',
            kpts=kgrid.calc_kpt_tuple(atoms))
atoms.set_calculator(calc)
atoms.get_total_energy()
```

would perform a VASP calculation in the current directory with the PBE
functional, using **kgrid** to determine the reciprocal-space sampling.

### kgrid-series

``` bash
kgrid-series FILE -t TYPE --min MIN --max MAX
```

Example output:

``` bash
Length cutoff  KSPACING    Samples
-------------  --------  ------------
      10.630    0.2956     2   7   4
      11.260    0.2790     2   8   4
      11.860    0.2649     2   8   5
      12.148    0.2586     3   8   5
      13.666    0.2299     3   9   5
      14.075    0.2232     3  10   5
      15.185    0.2069     3  10   6
      16.703    0.1881     3  11   6
      16.890    0.1860     3  12   6
      17.790    0.1766     3  12   7
      18.222    0.1724     4  12   7
      19.705    0.1594     4  13   7
      19.741    0.1591     4  13   8
      21.259    0.1478     4  14   8
      22.520    0.1395     4  15   8
      22.777    0.1379     4  15   9
      23.720    0.1324     4  16   9
      24.296    0.1293     5  16   9
      25.335    0.1240     5  17   9
      25.814    0.1217     5  17  10
      27.333    0.1149     5  18  10
      28.150    0.1116     5  19  10
      28.852    0.1089     5  19  11
      29.650    0.1060     5  20  11

```

Use the `--castep` option to replace KSPACING with "MP SPACING", which
corresponds to the *KPOINTS_MP_SPACING* parameter in CASTEP
(i.e. divide by 2π).

INSTALLATION
------------

**kgrid** uses setuptools; from a reasonable healthy Python environment you can use

    pip install .

with the usual pip caveats:

- the `--user` flag is highly recommended and avoids the need for administrator privileges, but on a somewhat unhealthy Python installation the user packages location may not be on your paths yet.
- the `-e` flag creates an "editable" installation which links to this repository and enables easy updates with git.

**kgrid** is not developed on Windows but no problems are anticipated; the Anaconda Python distribution includes pip. We have had good experiences using the Windows subsystem for Linux (WSL), available on Windows 10.
On Mac OSX, the system Python does not include pip but there are various ways of getting a more complete distribution such as Homebrew or Anaconda.

Testing
-------

To run the unit tests, install `pytest` and `pytest-mock` and run
`pytest` from the project directory (i.e. the folder containing this
README.)

Disclaimer
----------

This program is not affiliated with ASE or any particular quantum chemistry code.
This program is made available under the GNU General Public License; you are free to modify and use the code, but do so at your own risk.

References
----------

[1] Moreno, J., & Soler, J. (1992). Optimal meshes for integrals in real- and reciprocal-space unit cells. *Physical Review B*, 45(24), 13891–13898. [doi:10.1103/PhysRevB.45.13891](http://dx.doi.org/10.1103/PhysRevB.45.13891)


CHANGELOG.md
# Change Log

Notable changes are logged here by release. This project is not
expected to be especially active and follows a simplified Semantic Versioning:

- Version numbers take the format X.Y
- X is associated with major API breakage / changes in algorithm and results.
- Y is associated with minor updates and improvements
- Small amounts of code tidying, refactoring and documentation do not
  lead to a new release, and simply sit on the Development branch of
  the Git repository.

The changelog format is inspired by [keep-a-changelog](https://github.com/olivierlacan/keep-a-changelog).

## [Unreleased]

## [1.2] - 2021-04-01
- ASE deprecation warning cleared
- Minimum ASE version specified in setup.py
- This also means dropping support for Python 2.7
- Introduce unit testing

### Fixed
- ASE > 3.21 compatibility, backwards-compatible to ASE 3.18.

### Changed
- Atoms method (`atoms.get_reciprocal_cell()`) replaced by Cell method (`atoms.cell.reciprocal()`).

## [1.1] - 2018-04-30

### Fixed
- Python3 compatibility
- Bug in decimal place tolerance option

### Added
- Comma-separated kgrid-series output option
- CASTEP-like reciprocal spacing cutoff (2 pi factor smaller than KSPACING)

### Changed
- Make filename a positional argument; -f or --filename no longer needed

## [1.0] - 2016-10-07

### Changed
- Correctly calculate reciprocal lattice vectors using whole
  cell. This is the new default behaviour. Former behaviour ("naive"
  apprach using only lengths of real-space cell) made available as
  option through Python interface and CLI.
- Restructure repository for packaging with setuptools
- Installation instructions in README

### Added
- New emulation of VASP Auto and KSPACING modes.
- Setuptools installation
- New range generator for convergence testing: **kgrid-series**

## [0.2] - 2016-07-28

### Added
- Present useful Python API for use with [Atomic Simulation Environment (ASE)](https://wiki.fysik.dtu.dk/ase/) calculators

### Changed
- Import routine uses [Atomic Simulation Environment](https://wiki.fysik.dtu.dk/ase/). This adds a
  dependency, but permits import of almost any major crystal structure
  file format.

## 0.1 - 2013-06-01

### Added
- Basic codebase reading from FHI-aims input files
- Optparse-based interface
- GPL

[Unreleased]: https://github.com/wmd-group/kgrid/compare/v1.1...HEAD
[1.1]: https://github.com/wmd-group/kgrid/compare/v1.0...v1.1
[1.0]: https://github.com/wmd-group/kgrid/compare/v0.2...v1.0
[0.2]: https://github.com/wmd-group/kgrid/compare/v0.1...v0.2


test/test_kgrid_cli.py
import pathlib

import pytest

import kgrid.cli

data_dir = pathlib.Path('test/data')
sio2 = data_dir / 'POSCAR'


def test_calc_grid_output(capsys):
    kgrid.cli.calc_grid(15., filename=str(sio2), pretty_print=True)
    captured = capsys.readouterr()

    assert captured.out == '  8   8  12\n'


def test_calc_grid_format(tmp_path):
    tmpdir = tmp_path / "calc-grid"
    tmpdir.mkdir()

    with open(sio2) as fd:
        poscar_data = fd.read()

    (tmpdir / 'ambiguous-name').write_text(poscar_data)

    assert kgrid.cli.calc_grid(15., filename=(tmpdir / 'ambiguous-name'),
                               filetype='vasp') == (8, 8, 12)


@pytest.mark.parametrize('args, expected_args, expected_kwargs',
                         [(['the_file', '-a', '12'],
                           [12], {'filename': 'the_file',
                                  'mode': 'vasp_auto',
                                  'filetype': False,
                                  'realspace': False,
                                  'pretty_print': True}),
                          (['the_cell_file', '--castep_spacing', '0.05',
                            '-t', 'castep'],
                           [0.05], {'filename': 'the_cell_file',
                                    'mode': 'castep_mp_spacing',
                                    'filetype': 'castep',
                                    'realspace': False,
                                    'pretty_print': True}),
                          (['the_file', '-r'],
                           [10], {'filename': 'the_file',
                                  'mode': 'default',
                                  'filetype': False,
                                  'realspace': True,
                                  'pretty_print': True}),
                          ])
def test_kgrid_main(mocker, args, expected_args, expected_kwargs):
    mocked_calc_grid = mocker.patch('kgrid.cli.calc_grid',
                                    return_value=None)

    kgrid.cli.main(params=args)

    mocked_calc_grid.assert_called_with(*expected_args, **expected_kwargs)


test/test_kgrid_lib.py
import pathlib

import pytest
import ase.io

import kgrid

data_dir = pathlib.Path('test/data')
sio2 = ase.io.read(data_dir / 'POSCAR')
ga2o3 = ase.io.read(data_dir / 'geometry.in')


@pytest.mark.parametrize('atoms, expected, kwargs',
                         [(sio2, (5, 5, 8), {}),
                          (ga2o3, (2, 7, 4), {}),
                          (sio2, (2, 2, 4), {'mode': 'vasp_auto'}),
                          (sio2, (5, 5, 8), {'mode': 'kspacing',
                                             'cutoff_length': 0.3}),
                          (sio2, (5, 5, 8), {'mode': 'castep_mp_spacing',
                                             'cutoff_length': 0.05})])
def test_calc_kpt_tuple(atoms, expected, kwargs):
    assert kgrid.calc_kpt_tuple(atoms, **kwargs) == expected


test/test_kgrid_series_cli.py
import pathlib

from numpy.testing import assert_almost_equal
import ase
import ase.io
import pytest

import kgrid.series

data_dir = pathlib.Path('test/data')
sio2_file = data_dir / 'POSCAR'
sio2_atoms = ase.io.read(sio2_file)

nearly_cubic_atoms = ase.Atoms('Ag', cell=[1.0, 1.01, 1.001])

def test_get_increments():
    assert kgrid.series.get_increments([1, 2, 5]) == (0.5, 0.25, 0.1)


@pytest.mark.parametrize('args, kwargs, expected',
                         [([sio2_atoms, 3., 9.], {},
                           [4.0333, 4.2267, 5.3777, 6.34,
                             6.7221, 8.0665, 8.4533]),
                          ([nearly_cubic_atoms, 20, 21], {},
                           [20., 20.02, 20.2, 20.5, 20.5205, 20.705]),
                          ([nearly_cubic_atoms, 20, 21], {'decimals': 1},
                           [20., 20.2, 20.5, 20.7])
                         ])
def test_cutoff_series(args, kwargs, expected):
    result = kgrid.series.cutoff_series(*args, **kwargs)
    assert_almost_equal(result, expected)


@pytest.mark.parametrize('params, expected',
                         [([str(sio2_file), '-t', 'vasp',
                            '--min=4', '--max', '7',
                            '--comma_sep'],
                            "2 2 3,2 2 4,3 3 4,3 3 5,4 4 5\n"),
                          ([str(sio2_file),
                            '--min=2', '--max', '4',
                            '--castep'],
                           ('Length cutoff  MP SPACING    Samples\n'
                            '-------------  ----------  ------------\n'
                            '       2.113    0.236597     1   1   2\n'
                            '       2.689    0.185957     2   2   2\n'
                            )),])
def test_cli_output(capsys, params, expected):
    kgrid.series.main(params)
    captured = capsys.readouterr()

    assert captured.out == expected


kgrid.py
#! /usr/bin/env python

"""Get k-grid parameters for desired length cutoff and input geometry"""

###############################################################################
# Copyright 2015 Adam Jackson
###############################################################################
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
###############################################################################

import numpy as np
from argparse import ArgumentParser
import ase.io

def calc_grid(cutoff_length, mode='default', filename='geometry.in',
              filetype=False, realspace=False, pretty_print=False):

    if filetype:
        atoms = ase.io.read(filename, format=filetype)
    else:
        atoms = ase.io.read(filename)

    k_samples = calc_kpt_tuple(atoms, mode=mode,
                               cutoff_length=cutoff_length,
                               realspace=realspace)

    # Print vectors
    if pretty_print:
        print '{0:3.0f} {1:3.0f} {2:3.0f}'.format(*k_samples)
    else:
        return k_samples


def calc_kpt_tuple(atoms, cutoff_length=10, realspace=False, mode='default'):
    """Calculate k-point mesh from ASE atoms object for real-space cutoff

    Args:
        :param atoms: Periodic crystal structure
        :type ase.atoms.Atoms
        :param cutoff_length: Parameter determining sample density
        :type float
        :param mode: Calculation mode; this determines the units of
            cutoff_length.
            default: Cutoff is length in Angstroms corresponding to
                non-overlapping radius in a hypothetical supercell
                (Moreno-Soler length cutoff).
            vasp_auto: Cutoff is length in Angstroms to be multiplied by
                reciprocal lattice -- equivalent to double the value used in
                default mode. This attempts to emulate the "fully automatic"
                option in a VASP k-points file.
            kspacing: Cutoff has units of reciprocal Angstroms and represents
                minimum sampling density in reciprocal space. Equivalent to
                1/(2*cutoff) in default mode. This attempts to emulate the
                KSPACING parameter in VASP.
        :type str
        :returns: Number of samples in each reciprocal lattice vector
        :rtype: 3-tuple
        """

    if mode.lower() == 'default':
        rounding = 'up'
    elif mode.lower() == 'vasp_auto':
        cutoff_length = (cutoff_length) / 2.
        rounding = 'nearest'
    elif mode.lower() == 'kspacing':
        cutoff_length = np.pi / cutoff_length
        rounding = 'up'

    if realspace:
        return calc_kpt_tuple_naive(atoms, cutoff_length=cutoff_length,
                                    rounding=rounding)
    else:
        return calc_kpt_tuple_recip(atoms, cutoff_length=cutoff_length,
                                    rounding=rounding)


def calc_kpt_tuple_naive(atoms, cutoff_length=10, rounding='up'):
    """Calculate k-point grid using real-space lattice vectors"""

    # Import lattice vectors using ASE
    lattice_vectors = atoms.cell

    # Get lattice vector magnitudes (a, b, c) according to Pythagoras' theorem
    abc = np.sqrt(np.sum(np.square(lattice_vectors),1))

    # k-point samples required = 2*cutoff_length/(magnitude of lattice vector)
    k_samples = np.divide(2*cutoff_length,abc)

    # Rounding
    if rounding == 'up':
        k_samples = np.ceil(k_samples)
    else:
        k_samples = np.floor(k_samples + 0.5)
    return tuple((int(x) for x in k_samples))


def calc_kpt_tuple_recip(atoms, cutoff_length=10, rounding='up'):
    """Calculate reciprocal-space sampling with real-space parameter"""
    # Get reciprocal lattice vectors with ASE. Note that ASE does NOT include
    # the 2*pi factor used in many definitions of these vectors; the underlying
    # method is just a matrix inversoin and transposition
    recip_cell = atoms.get_reciprocal_cell()

    # Get reciprocal cell vector magnitudes according to Pythagoras' theorem
    abc_recip = np.sqrt(np.sum(np.square(recip_cell),1))

    k_samples = abc_recip * 2 * cutoff_length

    # Rounding
    if rounding == 'up':
        k_samples = np.ceil(k_samples)
    else:
        k_samples = np.floor(k_samples + 0.5)
    return tuple((int(x) for x in k_samples))

if __name__ == '__main__':
    parser = ArgumentParser()
    threshold = parser.add_mutually_exclusive_group()
    threshold.add_argument("-c", "--cutoff-length",
                      action="store", type=float, dest="cutoff_length",
                      default=10.0,
                      help="Set length cutoff in Angstroms [default: 10]")
    threshold.add_argument("-a", "--vasp-auto", action="store", type=float,
                      dest="vasp_auto",
                      help="Real-space cutoff like Auto in VASP KPOINTS file")
    threshold.add_argument("-s", "--vasp-kspacing",
                      action="store", type=float, dest="kspacing",
                      help="Reciprocal space distance like KSPACING in VASP")

    parser.add_argument("-f", "--file",
                      action="store", type=str, dest="file",
                      default="geometry.in",
                      help="Path to input file [default: ./geometry.in]")
    parser.add_argument("-t", "--type", action="store", type=str,
                      default=False,
                      help="Input file type. If not provided, ASE will guess.")
    parser.add_argument("-r", "--realspace", action="store_true",
                      help="Use real-space vector lengths instead of "
                           "computing reciprocal cell; not recommended!")
    # Add further options here
    args = parser.parse_args()

    if args.vasp_auto:
        mode = 'vasp_auto'
        cutoff = args.vasp_auto
    elif args.kspacing:
        mode = 'kspacing'
        cutoff = args.kspacing
    else:
        mode = 'default'
        cutoff = args.cutoff_length

    calc_grid(cutoff, mode=mode, filename=args.file,
              filetype=args.type, realspace=args.realspace,
              pretty_print=True)


kgrid/cli.py
#! /usr/bin/env python
###############################################################################
# Copyright 2016 Adam Jackson
###############################################################################
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
###############################################################################

from __future__ import print_function
import ase.io
from argparse import ArgumentParser
from kgrid import calc_kpt_tuple

def calc_grid(cutoff_length,
              mode='default',
              filename='geometry.in',
              filetype=False,
              realspace=False,
              pretty_print=False):

    if filetype:
        atoms = ase.io.read(filename, format=filetype)
    else:
        atoms = ase.io.read(filename)

    k_samples = calc_kpt_tuple(
        atoms, mode=mode, cutoff_length=cutoff_length, realspace=realspace)

    # Print vectors
    if pretty_print:
        print('{0:3.0f} {1:3.0f} {2:3.0f}'.format(*k_samples))
    else:
        return k_samples


def get_parser():
    parser = ArgumentParser()
    parser.add_argument(
        action="store",
        nargs="?",
        type=str,
        dest="file",
        default="geometry.in",
        help="Path to input file [default: ./geometry.in]")
    threshold = parser.add_mutually_exclusive_group()
    threshold.add_argument(
        "-c",
        "--cutoff-length",
        action="store",
        type=float,
        dest="cutoff_length",
        default=10.0,
        help="Set length cutoff in Angstroms [default: 10]")
    threshold.add_argument(
        "-a",
        "--vasp-auto",
        action="store",
        type=float,
        dest="vasp_auto",
        help="Real-space cutoff like Auto in VASP KPOINTS file")
    threshold.add_argument(
        "-s",
        "--vasp-kspacing",
        action="store",
        type=float,
        dest="kspacing",
        help="Reciprocal-space distance like KSPACING in VASP")
    threshold.add_argument(
        "--castep",
        "--castep_spacing",
        "--castep_mp_spacing",
        action="store",
        type=float,
        dest="castep_mp_spacing",
        help=("Reciprocal-space distance like KPOINTS_MP_SPACING in CASTEP; "
              "this differs from Vasp-like KSPACING by factor of 1/(2 pi)."))
    parser.add_argument(
        "-t",
        "--type",
        action="store",
        type=str,
        default=False,
        help="Input file type. If not provided, ASE will guess.")
    parser.add_argument(
        "-r",
        "--realspace",
        action="store_true",
        help="Use real-space vector lengths instead of "
        "computing reciprocal cell; not recommended!")

    return parser


def main(params=None):
    parser = get_parser()
    args = parser.parse_args(params)

    if args.vasp_auto:
        mode = 'vasp_auto'
        cutoff = args.vasp_auto
    elif args.kspacing:
        mode = 'kspacing'
        cutoff = args.kspacing
    elif args.castep_mp_spacing:
        mode = 'castep_mp_spacing'
        cutoff = args.castep_mp_spacing
    else:
        mode = 'default'
        cutoff = args.cutoff_length

    calc_grid(
        cutoff,
        mode=mode,
        filename=args.file,
        filetype=args.type,
        realspace=args.realspace,
        pretty_print=True)


if __name__ == '__main__':
    main()


kgrid/__init__.py
###############################################################################
# Copyright 2016 Adam Jackson
###############################################################################
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
###############################################################################

import numpy as np

def calc_kpt_tuple(atoms, cutoff_length=10, realspace=False, mode='default'):
    """Calculate k-point mesh from ASE atoms object for real-space cutoff

    Args:
        :param atoms: Periodic crystal structure
        :type ase.atoms.Atoms
        :param cutoff_length: Parameter determining sample density
        :type float
        :param mode: Calculation mode; this determines the units of
            cutoff_length.
            default: Cutoff is length in Angstroms corresponding to
                non-overlapping radius in a hypothetical supercell
                (Moreno-Soler length cutoff).
            vasp_auto: Cutoff is length in Angstroms to be multiplied by
                reciprocal lattice -- equivalent to double the value used in
                default mode. This attempts to emulate the "fully automatic"
                option in a VASP k-points file.
            kspacing: Cutoff has units of reciprocal Angstroms and represents
                minimum sampling density in reciprocal space. Equivalent to
                1/(2*cutoff) in default mode. This attempts to emulate the
                KSPACING parameter in VASP.
        :type str
        :returns: Number of samples in each reciprocal lattice vector
        :rtype: 3-tuple
        """

    if mode.lower() == 'default':
        rounding = 'up'
    elif mode.lower() == 'vasp_auto':
        cutoff_length = (cutoff_length) / 2.
        rounding = 'nearest'
    elif mode.lower() == 'kspacing':
        cutoff_length = np.pi / cutoff_length
        rounding = 'up'
    elif mode.lower() == 'castep_mp_spacing':
        cutoff_length = 1 / (2 * cutoff_length)
        rounding = 'up'

    if realspace:
        return calc_kpt_tuple_naive(atoms, cutoff_length=cutoff_length,
                                    rounding=rounding)
    else:
        return calc_kpt_tuple_recip(atoms, cutoff_length=cutoff_length,
                                    rounding=rounding)


def calc_kpt_tuple_naive(atoms, cutoff_length=10, rounding='up'):
    """Calculate k-point grid using real-space lattice vectors"""

    # Import lattice vectors using ASE
    lattice_vectors = atoms.cell

    # Get lattice vector magnitudes (a, b, c) according to Pythagoras' theorem
    abc = np.sqrt(np.sum(np.square(lattice_vectors),1))

    # k-point samples required = 2*cutoff_length/(magnitude of lattice vector)
    k_samples = np.divide(2*cutoff_length,abc)

    # Rounding
    if rounding == 'up':
        k_samples = np.ceil(k_samples)
    else:
        k_samples = np.floor(k_samples + 0.5)
    return tuple((int(x) for x in k_samples))


def calc_kpt_tuple_recip(atoms, cutoff_length=10, rounding='up'):
    """Calculate reciprocal-space sampling with real-space parameter"""
    # Get reciprocal lattice vectors with ASE. Note that ASE does NOT include
    # the 2*pi factor used in many definitions of these vectors; the underlying
    # method is just a matrix inversoin and transposition
    recip_cell = atoms.cell.reciprocal()

    # Get reciprocal cell vector magnitudes according to Pythagoras' theorem
    abc_recip = np.sqrt(np.sum(np.square(recip_cell),1))

    k_samples = abc_recip * 2 * cutoff_length

    # Rounding
    if rounding == 'up':
        k_samples = np.ceil(k_samples)
    else:
        k_samples = np.floor(k_samples + 0.5)
    return tuple((int(x) for x in k_samples))



kgrid/series.py
#! /usr/bin/env python

###############################################################################
# Copyright 2016 Adam Jackson
###############################################################################
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
###############################################################################
from __future__ import print_function

import numpy as np
from argparse import ArgumentParser

import ase.io

from kgrid import calc_kpt_tuple

def get_increments(lattice_lengths):
    """
    Calculate the vector l0 of increments between significant length cutoffs
    for each reciprocal lattice vector.

    :param lattice_lengths: Lengths of reciprocal lattice vectors
    :type 3-tuple

    :returns: Vector l0 of significant increments
    :rtype: 3-tuple
    """

    return tuple([1. / (2 * a) for a in lattice_lengths])


def cutoff_series(atoms, l_min, l_max, decimals=4):
    """Find multiples of l0 members within a range

    :param atoms: Crystal structure
    :type ase.atoms.Atoms
    :param l_min: Minimum real-space cutoff
    :type float
    :param l_max: Maximum real-space cutoff
    :type float
    :param decimals: Number of decimal places used when rounding to remove
        duplicates
    :type int

    :returns: Sorted list of cutoffs
    :rtype: list
    """
    recip_cell = atoms.cell.reciprocal()
    lattice_lengths = np.sqrt(np.sum(np.square(recip_cell), 1))

    l0 = get_increments(lattice_lengths)

    members = set()
    for li in l0:
        n_min = np.ceil(l_min / li)
        members.update(
            set(np.around(
                np.arange(n_min * li, l_max, li), decimals=decimals)))
    return sorted(members)


def kspacing_series(atoms, l_min, l_max, decimals=4):
    """Find series of KSPACING values with different results

    NB: It is strongly recommended to ADD a small delta to these values
    to account for truncation/rounding errors

    :param atoms: Crystal structure
    :type ase.atoms.Atoms
    :param l_min: Minimum real-space cutoff
    :type float
    :param l_max: Maximum real-space cutoff
    :type float

    :returns: Sorted list of KSPACING values
    :rtype: list
    """

    return [np.pi / c for c in
            cutoff_series(atoms, l_min, l_max, decimals=decimals)]

def get_parser():
    parser = ArgumentParser("Calculate a systematic series of k-point samples")
    parser.add_argument(
        'filename',
        nargs='?',
        type=str,
        default="geometry.in",
        help="Path to input file [default: ./geometry.in]")    
    parser.add_argument(
        '-t',
        '--type',
        type=str,
        default=None,
        help='Format of crystal structure file')
    parser.add_argument(
        '--min',
        type=float,
        default=10,
        help='Minimum real-space cutoff / angstroms')
    parser.add_argument(
        '--max',
        type=float,
        default=30,
        help='Maximum real-space cutoff / angstroms')
    parser.add_argument('--comma_sep', action='store_true',
                        help='Output as comma-separated list on one line')
    parser.add_argument('--castep', action='store_true',
                        help=('Provide CASTEP-like MP spacing instead of '
                              'vasp-like KSPACING'))
    return parser

def main(params=None):
    args = get_parser().parse_args(params)

    if args.type:
        atoms = ase.io.read(args.filename, format=args.type)
    else:
        atoms = ase.io.read(args.filename)

    cutoffs = cutoff_series(atoms, args.min, args.max)

    if args.castep:
        kspacing = [0.5 / c for c in cutoffs]
    else:
        kspacing = [np.pi / c for c in cutoffs]

    samples = [calc_kpt_tuple(
        atoms, cutoff_length=(cutoff - 1e-4)) for cutoff in cutoffs]

    if args.comma_sep:
        def print_sample(sample):
            return ' '.join((str(x) for x in sample))

        print(','.join((print_sample(sample) for sample in samples)))

    else:
        if args.castep:
            print("Length cutoff  MP SPACING    Samples")
            print("-------------  ----------  ------------")
            fstring = "{0:12.3f}   {1:9.6f}   {2:3d} {3:3d} {4:3d}"
        else:
            print("Length cutoff  KSPACING    Samples")
            print("-------------  --------  ------------")
            fstring = "{0:12.3f}   {1:7.4f}   {2:3d} {3:3d} {4:3d}"

        for cutoff, s, sample in zip(cutoffs, kspacing, samples):
            print(fstring.format(cutoff, s, *sample))


if __name__ == '__main__':
    main()
